### PR TITLE
Migrate k8s.io/csi-translation-lib/.* to structured logging

### LINKED ISSUE #120639: csi-translation-lib: Support structured and contextual logging
### What would you like to be added?

Migrate to use structured logging and contextual logging in csi-translation-lib.

### Why is this needed?

CSI Sidecar is advancing its support for structured logging and contextual logging. Therefore we need to implement support for structured logging and contextual logging in csi-translation-lib too, which is used internally by CSI Sidecar.

### PR DESCRIPTION
**Pull request by bells17**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Migrate to use structured logging and contextual logging in csi-translation-lib.
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md

I've implemented it to use contextual logging, as csi-translation-lib is a library and applications using csi-translation-lib may utilize contextual logging.
https://github.com/kubernetes-csi/csi-lib-utils/pull/149#pullrequestreview-1623868181

[The csi-translation-lib is used in the kubernetes-csi/external-provisioner](https://github.com/kubernetes-csi/external-provisioner/blob/d511a505621782e0924dba96285bd7b5f570ba49/go.mod#L26), and this modification is necessary to enable contextual logging within the kubernetes-csi/external-provisioner

based PR: https://github.com/kubernetes/kubernetes/pull/120640

#### Which issue(s) this PR fixes:

Fixes #120639

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go [hunk 1]
```diff
@@ -602,7 +609,8 @@ func TestEphemeralVolumeOwnerCheck(t *testing.T) {
 
 	podName := util.GetUniquePodName(pod)
 
-	dswp.findAndAddNewPods()
+	_, ctx := ktesting.NewTestContext(t)
```
Comment (carlory) [inline pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go:612]: Hi @pohly , Can you take a look? `NewTestContext(t)` and `Init(t)`, which is preferred?

Comment (pohly) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go:612]: `Init` is simpler and thus preferred.

Comment (pohly) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go:612]: Just beware that it cannot be used in staging (yet) - I first need to move k/k/test/utils/ktesting into a staging repo.

Comment (bells17) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go:612]: @carlory @pohly Thank you for your comments.

> I first need to move k/k/test/utils/ktesting into a staging repo.

Regarding this, is there anything I can help with, such as creating issues for repository creation at https://github.com/kubernetes/org?

**Comment by bells17** [comment]
/sig instrumentation
/wg structured-logging
/area logging

https://github.com/kubernetes/kubernetes/pull/120640#issuecomment-1718710578

**Comment by bells17** [comment]
/assign @mengjiao-liu

https://github.com/kubernetes/kubernetes/pull/120640#issuecomment-1729927135

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go (hunk 1 above)
Comment (pohly) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go:612]: I first need to polish it a bit more and gather more experience with it before proposing it for staging.

Comment (bells17) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go]: I changed to use `Init(t)` instead of `NewTestContext(t)`.
https://github.com/kubernetes/kubernetes/pull/124439/commits/aa67f41696d07aa5f333c6f2b7359548e074f267

**Comment by bells17** [comment]
/retitle Migrate k8s.io/csi-translation-lib/.* to structured logging
/wg structured-logging 
/area logging
/priority important-longterm
/kind cleanup
/cc @kubernetes/wg-structured-logging-reviews

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#what-to-include-in-the-pull-request

File: pkg/volume/util/util.go [hunk 2]
```diff
@@ -612,9 +612,9 @@ func GetLocalPersistentVolumeNodeNames(pv *v1.PersistentVolume) []string {
 // GetPodVolumeNames returns names of volumes that are used in a pod,
 // either as filesystem mount or raw block device, together with list
 // of all SELinux contexts of all containers that use the volumes.
-func GetPodVolumeNames(pod *v1.Pod) (mounts sets.String, devices sets.String, seLinuxContainerContexts map[string][]*v1.SELinuxOptions) {
-	mounts = sets.NewString()
-	devices = sets.NewString()
+func GetPodVolumeNames(pod *v1.Pod) (mounts sets.Set[string], devices sets.Set[string], seLinuxContainerContexts map[string][]*v1.SELinuxOptions) {
+	mounts = sets.New[string]()
+	devices = sets.New[string]()
```
Comment (mengjiao-liu) [inline pkg/volume/util/util.go:617]: Is it better to move this replacement commit into a separate PR?

File: staging/src/k8s.io/csi-translation-lib/plugins/azure_file.go [hunk 3]
```diff
@@ -59,21 +59,21 @@ func NewAzureFileCSITranslator() InTreePlugin {
...
+func (t *azureFileCSITranslator) TranslateInTreeInlineVolumeToCSI(logger klog.Logger, volume *v1.Volume, podNamespace string) (*v1.PersistentVolume, error) {
 	if volume == nil || volume.AzureFile == nil {
 		return nil, fmt.Errorf("volume is nil or Azure File not defined on volume")
 	}
 
 	azureSource := volume.AzureFile
 	accountName, err := getStorageAccountName(azureSource.SecretName)
 	if err != nil {
-		klog.V(5).Infof("getStorageAccountName(%s) returned with error: %v", azureSource.SecretName, err)
+		logger.Info("getStorageAccountName returned with error", "secretName", azureSource.SecretName, "err", err)
```
Comment (mengjiao-liu) [inline staging/src/k8s.io/csi-translation-lib/plugins/azure_file.go:76]: Why remove this log level？

File: staging/src/k8s.io/csi-translation-lib/plugins/azure_file.go [hunk 4]
```diff
@@ -112,15 +112,15 @@ func (t *azureFileCSITranslator) TranslateInTreeInlineVolumeToCSI(volume *v1.Vol
...
+func (t *azureFileCSITranslator) TranslateInTreePVToCSI(logger klog.Logger, pv *v1.PersistentVolume) (*v1.PersistentVolume, error) {
 	if pv == nil || pv.Spec.AzureFile == nil {
 		return nil, fmt.Errorf("pv is nil or Azure File source not defined on pv")
 	}
 
 	azureSource := pv.Spec.PersistentVolumeSource.AzureFile
 	accountName, err := getStorageAccountName(azureSource.SecretName)
 	if err != nil {
-		klog.V(5).Infof("getStorageAccountName(%s) returned with error: %v", azureSource.SecretName, err)
+		logger.Info("getStorageAccountName returned with error", "secretName", azureSource.SecretName, "err", err)
```
Comment (mengjiao-liu) [inline staging/src/k8s.io/csi-translation-lib/plugins/azure_file.go:123]: same above.

File: pkg/volume/util/util.go (hunk 2 above)
Comment (bells17) [reply pkg/volume/util/util.go:617]: Note to self:

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124439/pull-kubernetes-linter-hints/1788097670976376832

Comment (bells17) [reply pkg/volume/util/util.go:617]: I'm sorry, but it seems that we need to make a fix in this PR because the tests are failing due to the following lint error:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124439/pull-kubernetes-linter-hints/1788097670976376832

File: staging/src/k8s.io/csi-translation-lib/plugins/azure_file.go (hunk 3 above)
Comment (bells17) [reply staging/src/k8s.io/csi-translation-lib/plugins/azure_file.go]: Sorry,　I've fixed it。

File: staging/src/k8s.io/csi-translation-lib/plugins/azure_file.go (hunk 4 above)
Comment (bells17) [reply staging/src/k8s.io/csi-translation-lib/plugins/azure_file.go]: I've fixed it too.

**Comment by bells17** [comment]
@mengjiao-liu I've made the necessary corrections.
Would you please review it once more?

**Comment by mengjiao-liu** [comment]
image

Can you remove these commits to keep a clean git history? You can use the [git rebase](https://git-scm.com/docs/git-rebase) command.

**Comment by bells17** [comment]
@mengjiao-liu Sure, I've rebased the commits. Would you please review the changes again?

File: staging/src/k8s.io/csi-translation-lib/plugins/rbd.go [hunk 5]
```diff
@@ -0,0 +1,326 @@
+/*
```
Comment (carlory) [inline staging/src/k8s.io/csi-translation-lib/plugins/rbd.go:1]: The file was removed by https://github.com/kubernetes/kubernetes/pull/124546 Don't re-add it.

Comment (bells17) [reply staging/src/k8s.io/csi-translation-lib/plugins/rbd.go:1]: Sorry, I'll remove it.

Comment (bells17) [reply staging/src/k8s.io/csi-translation-lib/plugins/rbd.go]: Done.

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go [hunk 6]
```diff
@@ -50,7 +50,7 @@ import (
 // if it has volumes. It also verifies that the pods in the desired state of the
 // world cache still exist, if not, it removes them.
 type DesiredStateOfWorldPopulator interface {
-	Run(sourcesReady config.SourcesReady, stopCh <-chan struct{})
+	Run(ctx context.Context, sourcesReady config.SourcesReady, stopCh <-chan struct{})
```
Comment (carlory) [inline pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:53]: Can we use `ctx.Done()` instead?

```suggestion
	Run(ctx context.Context, sourcesReady config.SourcesReady)
```

Comment (bells17) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:53]: Thank you. I'll fix it.

Comment (bells17) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:53]: @carlory Excuse me, I hope you don't mind me asking a question.

Currently, `wait.NeverStop` is being passed from kubelet to `stopCh`. However, if we change it to `ctx.Done()`, it will be replaced with the one from `context.Background()`.
Would this change cause any issues, or is it acceptable?

- https://github.com/kubernetes/kubernetes/blob/2f3c8e63ab3a89e3930f7f8611c7e27d06f06045/pkg/kubelet/kubelet.go#L1633
-  https://github.com/kubernetes/kubernetes/blob/2f3c8e63ab3a89e3930f7f8611c7e27d06f06045/pkg/kubelet/volumemanager/volume_manager.go#L286

Comment (carlory) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:53]: I think it makes sense. i.e.

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1649

Comment (bells17) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:53]: Thank you for your response. If you don't mind, I would like to ask an additional question.

There are several other places within `*Kubelet.Run()` where `wait.NeverStop` is being used. Should we also update those `wait.NeverStop` to use `ctx.Done()` for consistency?

- https://github.com/kubernetes/kubernetes/blob/2f3c8e63ab3a89e3930f7f8611c7e27d06f06045/pkg/kubelet/kubelet.go#L1623
- https://github.com/kubernetes/kubernetes/blob/2f3c8e63ab3a89e3930f7f8611c7e27d06f06045/pkg/kubelet/kubelet.go#L1645
- https://github.com/kubernetes/kubernetes/blob/2f3c8e63ab3a89e3930f7f8611c7e27d06f06045/pkg/kubelet/kubelet.go#L1651
- https://github.com/kubernetes/kubernetes/blob/2f3c8e63ab3a89e3930f7f8611c7e27d06f06045/pkg/kubelet/kubelet.go#L1663

Comment (bells17) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:53]: @carlory Thank you for your response. If you don't mind, I would like to ask an additional question.

There are several other places within `*Kubelet.Run()` where `wait.NeverStop` is being used. Should we also update those `wait.NeverStop` to use `ctx.Done()` for consistency?

- https://github.com/kubernetes/kubernetes/blob/2f3c8e63ab3a89e3930f7f8611c7e27d06f06045/pkg/kubelet/kubelet.go#L1623
- https://github.com/kubernetes/kubernetes/blob/2f3c8e63ab3a89e3930f7f8611c7e27d06f06045/pkg/kubelet/kubelet.go#L1645
- https://github.com/kubernetes/kubernetes/blob/2f3c8e63ab3a89e3930f7f8611c7e27d06f06045/pkg/kubelet/kubelet.go#L1651
- https://github.com/kubernetes/kubernetes/blob/2f3c8e63ab3a89e3930f7f8611c7e27d06f06045/pkg/kubelet/kubelet.go#L1663

Comment (carlory) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:53]: If changing it to `Run(ctx context.Context, sourcesReady config.SourcesReady)` has lots of changes, let's do it step by step. this PR only adds the `ctx` parameter to the `Run` method.

**Comment by yangjunmyfm192085** [comment]
Hi, @bells17  Can you Squash the commits?

**Review by yangjunmyfm192085 — APPROVED** [review]
Hi, @bells17 Can you Squash the commits?

**Comment by bells17** [comment]
@yangjunmyfm192085 Thank you. I squashed the commits.

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go (hunk 6 above)
Comment (bells17) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go]: I changed `kl.volumeManager.Run()` to use `ctx.Done()` instead of `wait.NeverStop`.
Would you please review this again?

File: pkg/kubelet/kubelet_test.go [hunk 7]
```diff
@@ -2996,9 +2996,9 @@ func simulateVolumeInUseUpdate(
 	}
 }
 
-func runVolumeManager(kubelet *Kubelet) chan struct{} {
+func runVolumeManager(ctx context.Context, kubelet *Kubelet) chan struct{} {
```
Comment (carlory) [inline pkg/kubelet/kubelet_test.go:2999]: This function becomes useless. we don't need to get the channel.

File: pkg/kubelet/kubelet_volumes_test.go [hunk 8]
```diff
@@ -78,7 +79,8 @@ func TestListVolumesForPod(t *testing.T) {
 		},
 	})
 
-	stopCh := runVolumeManager(kubelet)
+	tCtx := ktesting.Init(t)
+	stopCh := runVolumeManager(tCtx, kubelet)
```
Comment (carlory) [inline pkg/kubelet/kubelet_volumes_test.go:83]: go kubelet.volumeManager.Run(ctx, kubelet.sourcesReady)

Comment (carlory) [inline pkg/kubelet/kubelet_volumes_test.go:83]:
```suggestion
	go kubelet.volumeManager.Run(ctx, kubelet.sourcesReady)
```

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go (hunk 6 above)
Comment (carlory) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go]: @bells17 some nits, otherwise LGTM

Comment (bells17) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go]: @carlory Thank you. I fixed the test cases.
Would you please review this again? I'll squash the commits if there are no problems with my commit.
https://github.com/kubernetes/kubernetes/pull/124439/commits/728bed6f5ac266193750174992a458f57525ef80

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go [hunk 9]
```diff
@@ -141,21 +141,23 @@ type processedPods struct {
...
 
-func (dswp *desiredStateOfWorldPopulator) Run(sourcesReady config.SourcesReady, stopCh <-chan struct{}) {
+func (dswp *desiredStateOfWorldPopulator) Run(ctx context.Context, sourcesReady config.SourcesReady) {
 	// Wait for the completion of a loop that started after sources are all ready, then set hasAddedPods accordingly
-	klog.InfoS("Desired state populator starts to run")
-	wait.PollUntil(dswp.loopSleepDuration, func() (bool, error) {
+	logger := klog.FromContext(ctx)
+	logger.Info("Desired state populator starts to run")
+	// nolint: errcheck
+	wait.PollUntilContextCancel(ctx, dswp.loopSleepDuration, false, func(ctx context.Context) (bool, error) {
```
Comment (carlory) [inline pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:149]: please check err, `nolint` isn't recommended.

Comment (carlory) [inline pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:149]: remove `// nolint: errcheck`
```suggestion
	_ = wait.PollUntilContextCancel(ctx, dswp.loopSleepDuration, false, func(ctx context.Context) (bool, error) {
```

Comment (carlory) [inline pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:149]: remove `// nolint: errcheck`
```suggestion
	_ = wait.PollUntilContextCancel(ctx, dswp.loopSleepDuration, false, func(ctx context.Context) (bool, error) {
```

Comment (carlory) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:149]: `_ = ` is more general in k/k

**Comment by carlory** [comment]
@bells17  Git Tips: It will be useful when you want to add some changes to the last commit without creating a new commit.

```sh
git add .
git commit -a --amend --no-edit
git push -f 
```

please squash your comments, I will lgtm it

**Comment by bells17** [comment]
@carlory I have squashed the commits, so I would appreciate it if you could review the changes.

> Git Tips: It will be useful when you want to add some changes to the last commit without creating a new commit.

Thank you. I will keep that in mind for reference.

**Comment by pohly** [comment]
/assign

For the hack directory approval.

**Comment by bells17** [comment]
/assign @jsafrane

For:

- pkg/controller/volume/OWNERS
- pkg/scheduler/framework/plugins/nodevolumelimits/OWNERS
- pkg/scheduler/framework/plugins/volumebinding/OWNERS
- pkg/volume/OWNERS
- staging/src/k8s.io/csi-translation-lib/OWNERS

**Comment by bells17** [comment]
/assign @klueska

For: pkg/kubelet/OWNERS

File: pkg/scheduler/framework/plugins/nodevolumelimits/csi.go [hunk 10]
```diff
@@ -45,8 +45,8 @@ type InTreeToCSITranslator interface {
 	IsMigratableIntreePluginByName(inTreePluginName string) bool
 	GetInTreePluginNameFromSpec(pv *v1.PersistentVolume, vol *v1.Volume) (string, error)
 	GetCSINameFromInTreeName(pluginName string) (string, error)
-	TranslateInTreePVToCSI(pv *v1.PersistentVolume) (*v1.PersistentVolume, error)
-	TranslateInTreeInlineVolumeToCSI(volume *v1.Volume, podNamespace string) (*v1.PersistentVolume, error)
+	TranslateInTreePVToCSI(logger klog.Logger, pv *v1.PersistentVolume) (*v1.PersistentVolume, error)
+	TranslateInTreeInlineVolumeToCSI(logger klog.Logger, volume *v1.Volume, podNamespace string) (*v1.PersistentVolume, error)
```
Comment (pohly) [inline pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:49]: We don't have any out-of-tree users of this API, right?

Otherwise a `ctx context.Context` parameter would be better, just in case that we ever need it. If it's in-tree, we can still switch later when the need arises.

/lgtm
/approve

For hack config update.

Comment (bells17) [reply pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:49]: Thank you.

> We don't have any out-of-tree users of this API, right?

To the best of my knowledge, there don't seem to be any instances where out-of-tree users, including CSI Sidecar and kube-scheduler-simulator, directly use `TranslateInTreePVToCSI` and `TranslateInTreeInlineVolumeToCSI` .

**Comment by pohly** [comment]
/assign @klueska 

For minor kubelet change approval.

**Review by mengjiao-liu — COMMENTED** [review]
/lgtm

**Comment by carlory** [comment]
/assign @derekwaynecarr @dchen1107

For minor kubelet change approval.

**Comment by dims** [comment]
@bells17 can you please rebase this to current master?

**Comment by mengjiao-liu** [comment]
Flakying test: https://github.com/kubernetes/kubernetes/issues/125903

**Comment by jsturtevant** [comment]
We've made an update that should resolve that test
/test pull-kubernetes-e2e-capz-windows-master

**Comment by jsturtevant** [comment]
All the test passed but one which is tracked in https://github.com/kubernetes/kubernetes/issues/126180 due to a new feasted that got turned on today

**Comment by nojnhuh** [comment]
CAPZ tests should be fixed now. If not, this should fix them: https://github.com/kubernetes/test-infra/pull/33022

/retest

**Comment by pohly** [comment]
/assign @SergeyKanzhelev

For SIG Node approval.

**Comment by SergeyKanzhelev** [comment]
> Does this PR introduce a user-facing change?

Maybe add a release note here?

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

kubelet change looks ok
